### ISSUE TITLE
Paths with named wildcards must be unique across all hosts

### ISSUE THREAD

**Issue by herojan** [zalando/skipper#1076]
**Describe the bug**
(This was previously debugged with help from the Skipper team so I'll document the findings here)

Skipper builds a tree for its routes with path predicates at the top, and other predicates such as host matches afterwards. When two path predicates are the same but with different names for their wildcard segments, they are ambiguous and skipper will fail to apply one of the routes, even if they are from different hosts.

**To Reproduce**
Apply the following routes to Skipper

```
wildcard_host_1: Host(/^testapp[.]some-host[.]zalan[.]do$/) && Method("GET") && Path("/resources/:id")
  -> lifo(2000, 20000, "10s")
  -> oauthTokeninfoAllScope("uid")
  -> "http://10.2.205.6:8080";
```

Then apply 

```
wildcard_host_2: Host(/^testapp2[.]some-host[.]zalan[.]do$/) && Method("GET") && Path("/resources/:resource-id")
  -> lifo(2000, 20000, "10s")
  -> "http://10.2.205.6:8080";
```

**Expected behavior**

I expect that I can apply **wildcard_host_2** (its host is different) without any issues.

**Observed behavior**

I see the following error message

```
wildcards [id] are ambiguous with wildcards [abc]
```

and the **wildcard_host_2** is never applied. 

This is bad in a distributed environment with many hosts in particular, such as kubernetes, as different hosts can possibly have the same path and they may not even be aware another host with the same path exists. 

Developers may not realise their route has not been applied and this could easily be a security problem if for example an authentication route is not applied.

**Comment by xavivars** [comment]
This will actually force us to have a different skipper deployment per expected host, due to the possible (probable) route collision. Is there any way to have a "pre-filter" based on hosts, or specify somehow that the host filter has preference?

**Comment by szuecs** [comment]
@xavivars right now I don’t see a workaround besides build your own skipper proxy (library first makes it possible), but I think we see the problem and will fix it soon.

Thanks for providing feedback on this issue

**Comment by xavivars** [comment]
I'm already using Skipper as a library (to be able to add filters/dataclients), so I'd be interested in knowing how that workaround could look like. Also, if there's anything I can can do to help fixing it, I'm willing to help

**Comment by szuecs** [comment]
I think there’s a more simple solution than I thought before. You would need to use `PathRegexp` instead of `Path` or `PathSubtree`. In Kubernetes Ingress you can use https://opensource.zalando.com/skipper/kubernetes/ingress-usage/#ingress-path-handling to choose it.

Why this should work you can read in https://opensource.zalando.com/skipper/reference/architecture/#route-matching

I hope this helps.

**Comment by cjbooms** [comment]
Our team's log monitoring just alerted us to the occurrence of this issue in our K8S cluster. 
Is it possible to get this bug triaged and worked on in the immediate future?
I worry that there is an increasing probability of this bug occurring in a clusters that my team cannot monitor the logs for.

**Comment by xavivars** [comment]
Just to make sure: this is only happening with named wildcards? Or is this happening with other use cases (not named wildcards, paths without wildcards,...)?

**Comment by szuecs** [comment]
@xavivars I am not sure what you refer with "this", maybe you can clarify what you exactly mean. Thanks

**Comment by NorthFury** [comment]
will be fixed by #1227

**Comment by xavivars** [comment]
> @xavivars I am not sure what you refer with "this", maybe you can clarify what you exactly mean. Thanks

You're right... my bad... I meant "the problem this issue is about (_Paths with named wildcards must be unique across all hosts_)"

**Comment by NorthFury** [comment]
fixed by #1227
